File: solutions/javascript/bob/1/bob.js
class BobMessages {
  static QUESTION = "Sure.";
  static YELL = "Whoa, chill out!";
  static YELLED_QUESTION = "Calm down, I know what I'm doing!";
  static SILENCE = "Fine. Be that way!";
  static DEFAULT = "Whatever.";
}

function isQuestion(message) {
  return /\?+\s*$/.test(message);
}

function isYelling(message) {
  return /^[^a-z]*[A-Z][^a-z]*$/.test(message);
}

function isSilence(message) {
  return /^\s*$/.test(message);
}

export const hey = (message) => {
  const silence = isSilence(message);
  if (silence) return BobMessages.SILENCE;
  
  const question = isQuestion(message);
  const yell = isYelling(message);
  if (question && yell) return BobMessages.YELLED_QUESTION;
  if (question) return BobMessages.QUESTION;
  if (yell) return BobMessages.YELL;
  
  return BobMessages.DEFAULT;
};

File: solutions/cpp/bob/1/bob.cpp
#include "bob.h"

namespace bob {
    namespace message {
        enum kind : uint8_t {
            question,
            yelling,
            yelled_question,
            silence,
            other
        };
    }    

    constexpr std::array<const char*, 5> response = {
        "Sure.",
        "Whoa, chill out!",
        "Calm down, I know what I'm doing!",
        "Fine. Be that way!",
        "Whatever."    
    };

    auto find_response(message::kind dialogue) -> std::string {
        return response[dialogue];
    }

    auto hey(const std::string& dialogue) -> std::string {
        if (const auto silence = std::all_of(dialogue.begin(), dialogue.end(),
            [](unsigned char character) -> bool {
                return std::isspace(character);
            }); silence) {
            return find_response(message::kind::silence);
        }

        const auto yell = std::any_of(dialogue.begin(), dialogue.end(),
            [](unsigned char character) {
                return std::isalpha(character);
            }) and 
            std::all_of(dialogue.begin(), dialogue.end(),
            [](unsigned char character) -> bool {
                if (std::isalpha(character)) {
                    return std::isupper(character);
                }
                return true;
            });

        const auto question = *(std::find_if(dialogue.rbegin(), dialogue.rend(),
            [](unsigned char character) -> bool {
                return not isspace(character);
            })) == '?'; 

        if (yell and question) {
            return find_response(message::kind::yelled_question);
        }
        else if (yell) {
            return find_response(message::kind::yelling);
        }
        else if (question) {
            return find_response(message::kind::question);
        }
              
        return find_response(message::kind::other);
    }
}  // namespace bob
